### PR TITLE
feat(datasource/custom): Support digest in custom datasource

### PR DESCRIPTION
**Pull request by fstehle**
## Changes

To support digests a datasource has to implement the `getDigest` method. 
This PR implements the method returning a null promise.
The value itself can be provided in the getReleases method.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by secustor — CHANGES_REQUESTED** [review]
Not fond of this solution, tough a better fix would need bigger changes.

**Comment by viceice** [comment]
why do we need this? that function shouldn't be called when the digest is already returned in releases.

**Comment by fstehle** [comment]
> Not fond of this solution, tough a better fix would need bigger changes.

Yes, absolutely. Would you suggest to modify `supportsDigests` in https://github.com/renovatebot/renovate/blob/6884d83a0573ebd40210cf0717c1814afd68a649/lib/modules/datasource/index.ts#L392 instead?

**Comment by fstehle** [comment]
> why do we need this? that function shouldn't be called when the digest is already returned in releases.

That is not what I observed, `supportsDigests` returns false and no digest update is made.

See the following lines for that:
* https://github.com/renovatebot/renovate/blob/6884d83a0573ebd40210cf0717c1814afd68a649/lib/modules/datasource/index.ts#L392)
* https://github.com/renovatebot/renovate/blob/6884d83a0573ebd40210cf0717c1814afd68a649/lib/workers/repository/process/lookup/index.ts#L417-L545

File: lib/modules/datasource/custom/index.ts
Comment (secustor) [inline lib/modules/datasource/custom/index.ts:68]:
```suggestion
    // istanbul ignore next
    return Promise.resolve(null)
```

File: lib/workers/repository/process/lookup/__snapshots__/index.spec.ts.snap
Comment (secustor) [inline lib/workers/repository/process/lookup/__snapshots__/index.spec.ts.snap:1]: The snapshot is with the latest change obsolete. 
You can remove it with `pnpm run jest -u`

**Review by secustor — CHANGES_REQUESTED** [review]

**Comment by secustor** [comment]
> > Not fond of this solution, tough a better fix would need bigger changes.
> 
> Yes, absolutely. Would you suggest to modify `supportsDigests` in
> 
> https://github.com/renovatebot/renovate/blob/6884d83a0573ebd40210cf0717c1814afd68a649/lib/modules/datasource/index.ts#L392
> instead?

The logic in `lookup/index.ts` would need to modified, which needs extensive testing on real repos.

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs deconflicting and continuing

**Comment by acortelyou** [comment]
This PR would resolve request for help discussion #28343.

I hope we can find a way to get this code moving again.

It looks like the suggestions from the review were just nits: if @fstehle doesn't have bandwidth to revisit, perhaps the maintainers can update the PR with their fixes or an interested party (myself) could fork the PR.

**Comment by rarkins** [comment]
The latter

**Comment by fstehle** [comment]
Currently quite busy, would be great if you could create a fork PR

**Review by rarkins — COMMENTED** [review]
Fixed conflicts, but now I defer to @secustor

File: lib/modules/datasource/custom/index.ts
Comment (viceice) [inline lib/modules/datasource/custom/index.ts:66]: should be removed, it seems there's a test for it

**Review by secustor — COMMENTED** [review]
Needs test fix and removal of ignore statement, else LGTM

**Comment by acortelyou** [comment]
See linked PR for commits to address outstanding items.

Either cherry-pick the commits on to here (i don't have permissions) or publish that one and merge from there.

Thanks!

File: lib/modules/datasource/custom/index.ts
Comment (acortelyou) [inline lib/modules/datasource/custom/index.ts:66]:
```suggestion
```

**Review by acortelyou — APPROVED** [review]

**Comment by ivankatliarchuk** [comment]
Does anyone know whether or not these changes will support Gitlab MD5 https://gitlab-docker-machine-downloads.s3.amazonaws.com/v0.16.2-gitlab.25/index.html?

**Comment by secustor** [comment]
> Does anyone know whether or not these changes will support Gitlab MD5 https://gitlab-docker-machine-downloads.s3.amazonaws.com/v0.16.2-gitlab.25/index.html?

No.

Closing as this has been implemented in the linked PR.

**Pull request closed without merging**
